### ISSUE TITLE
Only restore changes

### ISSUE THREAD

**Issue by MadisonCourto** [MySqlBackupNET/MySqlBackup.Net#39]
I have an remote DB that I need to take a backup of every hour then restore a the local DB with the remote data, this may be outside of scope, but would is it possible to only restore the changes rather than the whole database, currently if I do a restore all the data is dropped and I don't have access to the data for 15 minutes or so every hour.

I think I'm potentially looking at this wrong and want to do a differential database backup.

Cheers.

**Comment by adriancs2** [comment]
below is what I came out in the first thought:

Save the date time of every data export action taken from remote DB >> "LastBackupTime"
Then every rows in all table will have "last update time" > "LastUpdateTime".

Whenever you perform the next data export from remote DB, only get new data by comparing the "LastUpdateTime" to "LastBackupTime". All rows with "LastUpdateTime" older than "LastBackupTime" will be exported.

At MySqlBackup.NET part, it will be something like this:

```C#
DateTime LastBackuptime = GetLastBackupTime();

var dic = new Dictionary<string, string>();

dic["tableA"] = string.Format("SELECT * FROM `tableA` where last_update_time>'{0:yyyy-MM-dd HH:mm:ss}';", LastBackuptime);
dic["tableB"] = string.Format("SELECT * FROM `tableB` where last_update_time>'{0:yyyy-MM-dd HH:mm:ss}';", LastBackuptime);
dic["tableC"] = string.Format("SELECT * FROM `tableC` where last_update_time>'{0:yyyy-MM-dd HH:mm:ss}';", LastBackuptime);
dic["tableD"] = string.Format("SELECT * FROM `tableD` where last_update_time>'{0:yyyy-MM-dd HH:mm:ss}';", LastBackuptime);

using (MySqlConnection conn = new MySqlConnection(connstr))
{
    using (MySqlCommand cmd = new MySqlCommand())
    {
        using (MySqlBackup mb = new MySqlBackup(cmd))
        {
            cmd.Connection = conn;
            conn.Open();

            mb.ExportInfo.WrapWithinTransaction = true;
            mb.ExportInfo.RowsExportMode = RowsDataExportMode.Replace;
            mb.ExportInfo.TablesToBeExportedDic = dic;
            mb.ExportToFile("C:\\backup.sql");

            conn.Close();
        }
    }
}

DateTime newBackupTime = DateTime.UtcNow;

SaveNewBackupTime(newBackupTime);
```
Always use "DateTime.UtcNow" for recording the execution time, especially if both of your Databases are in different timezone.

**Comment by adriancs2** [comment]
```
mb.ExportInfo.RowsExportMode = RowsDataExportMode.Replace;
```
In stead of traditional "INSERT INTO", the data will be using "REPLACE INTO".
This works same as the "INSERT", instead, if the primary existed, the row will be deleted then "INSERT" with new data. This will not work on tables with **no primary key**.

**Comment by MadisonCourto** [comment]
Hey thanks for the reply, I have forked and been messing around with trying something similar to this making changes to reflect this https://stackoverflow.com/questions/16974120/mysql-import-only-missing-rows I have been looking at making changes to MySqlBackup.cs 

Line: 391 `Export_WriteLine(string.Format("DROP TABLE IF EXISTS `{0}`;", tableName));`
&
Line: 690  ` if (rowsExportMode == RowsDataExportMode.Insert)
                sb.Append("INSERT INTO `");
            else if (rowsExportMode == RowsDataExportMode.InsertIgnore)
                sb.Append("INSERT IGNORE INTO `");
            else if (rowsExportMode == RowsDataExportMode.Replace)
                sb.Append("REPLACE INTO `");`

Basically I think I'm on the right track, I'd like to get your thoughts, thanks!

**Comment by adriancs2** [comment]
What you mean is you don't want the "DROP TABLE" and "CREATE TABLE" statement in the dump file, and you wish to use the statement "INSERT IGNORE". Is it?

**Comment by MadisonCourto** [comment]
That’s correct, I think that will give me the indebted behaviour right?

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by adriancs2** [comment]
Then, you can apply 2 options to perform that:
```C#
mb.ExportInfo.ExportTableStructure = false;
mb.ExportInfo.RowsExportMode = RowsDataExportMode.InsertIgnore;
```
This line:
mb.ExportInfo.ExportTableStructure = false;
will skip and ignore "DROP TABLE" and "CREATE TABLE".

This line:
mb.ExportInfo.RowsExportMode = RowsDataExportMode.InsertIgnore;
In stead of the ordinary "INSERT INTO", now the dump will use "INSERT IGNORE"

**Comment by MadisonCourto** [comment]
Awesome thanks a lot mate!

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by adriancs2** [comment]
You're welcome.
